elasticdl/python/ps/embedding_table.py

```diff
@@ -44,34 +46,38 @@ def __init__(self, name, dim=None, initializer=None, is_slot=False):
             )
         self.is_slot = is_slot
         self.embedding_vectors = {}
+        self._lock = threading.Lock()
 
     def get(self, indices):
         if len(indices) == 0:
             return None
         values = []
-        for i in indices:
-            value = self.embedding_vectors.get(i, None)
-            if value is None:
-                value = self.initializer(shape=(self.dim,)).numpy()
-                self.embedding_vectors[i] = value
-            values.append(value)
+        with self._lock:
+            for i in indices:
+                value = self.embedding_vectors.get(i, None)
+                if value is None:
```

Comment:
Only adding embedding vector and iteration need the lock. Reading or updating existing vectors does not.
So Move the lock here after `if value is None:` for adding. 

Also, `def set(self, indices, values)` below  does not need the lock, as it does not add any vectors.

Reply:
We need a RWLock here. Please refer to the implementation in Go PS. https://github.com/sql-machine-learning/elasticdl/blob/develop/elasticdl/pkg/common/embedding_table.go#L14

The lock is used at `GetEmbeddingVector ` func and `ToIndexedSlices ` func.

Reply:
If we don't use lock the `def set(self, indices, values)`, Reading and writing the `embedding_vector` are simultaneous which may cause inconsistency.

Reply:
@workingloong  Yes. I am not sure if the reading/writeing concurrently will hurt the model accuracy. It seems that it's a tradeoff here, training speed and model accuracy. We could do some expriments.

Reply:
In async-SGD, reading/writing inconsistency does not matter, as the async-SGD itself has a staleness issue.
In sync-SGD, there won't be any reading/writing inconsistency.

Reply:
OK, we may need to do some experiments to view training speed and model accuracy and then to add the lock.